agents/community/open-deep-research-agent/open_deep_research_agent/server.py

```diff
@@ -0,0 +1,214 @@
+import argparse
+import asyncio
+import os
+import threading
+
+from typing import Any
+
+from acp.server.highlevel import Server
+from beeai_sdk.providers.agent import run_agent_provider
+from beeai_sdk.schemas.base import Log
+from beeai_sdk.schemas.metadata import Metadata, UiDefinition, UiType
+from beeai_sdk.schemas.text import TextInput, TextOutput
+
+from pydantic import Field
+from dotenv import load_dotenv
+from huggingface_hub import login
+from open_deep_research_agent.scripts.text_inspector_tool import TextInspectorTool
+from open_deep_research_agent.scripts.text_web_browser import (
+    ArchiveSearchTool,
+    FinderTool,
+    FindNextTool,
+    PageDownTool,
+    PageUpTool,
+    SimpleTextBrowser,
+    VisitTool,
+)
+from open_deep_research_agent.scripts.visual_qa import visualizer
+
+from smolagents import (
+    CodeAgent,
+    GoogleSearchTool,
+    # HfApiModel,
+    LiteLLMModel,
+    ToolCallingAgent,
+)
+
+
+AUTHORIZED_IMPORTS = [
+    "requests",
+    "zipfile",
+    "os",
+    "pandas",
+    "numpy",
+    "sympy",
+    "json",
+    "bs4",
+    "pubchempy",
+    "xml",
+    "yahoo_finance",
+    "Bio",
+    "sklearn",
+    "scipy",
+    "pydub",
+    "io",
+    "PIL",
+    "chess",
+    "PyPDF2",
+    "pptx",
+    "torch",
+    "datetime",
+    "fractions",
+    "csv",
+]
+load_dotenv(override=True)
+login(os.getenv("HF_TOKEN"))
+
+append_answer_lock = threading.Lock()
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "question",
+        type=str,
+        help="for example: 'How many studio albums did Mercedes Sosa release before 2007?'",
+    )
+    parser.add_argument("--model-id", type=str, default="o1")
+    return parser.parse_args()
+
+
+custom_role_conversions = {"tool-call": "assistant", "tool-response": "user"}
+
+user_agent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36 Edg/119.0.0.0"
+
+BROWSER_CONFIG = {
+    "viewport_size": 1024 * 5,
+    "downloads_folder": "downloads_folder",
+    "request_kwargs": {
+        "headers": {"User-Agent": user_agent},
+        "timeout": 300,
+    },
+    "serpapi_key": os.getenv("SERPAPI_API_KEY"),
+}
+
+os.makedirs(f"./{BROWSER_CONFIG['downloads_folder']}", exist_ok=True)
+
+
+def create_agent():
+    model_id = os.getenv("LITELLM_MODEL_ID", "ollama/llama3.1")
+    model_params = {
+        "api_base": os.getenv("LITELLM_API_BASE", "http://localhost:11434"),
+        "api_key": os.getenv("OPENAI_API_KEY", ""),
+        "model_id": model_id,
+        "custom_role_conversions": custom_role_conversions,
+        # "max_completion_tokens": 8192,
+    }
+    if model_id == "o1":
+        model_params["reasoning_effort"] = "high"
+    model = LiteLLMModel(**model_params)
+
+    text_limit = 100000
+    browser = SimpleTextBrowser(**BROWSER_CONFIG)
+    WEB_TOOLS = [
+        GoogleSearchTool(provider="serper"),
```

Comment:
Make this conditional (only include when API key env var is present).

Reply:
FWIW this key is required in the original (and I've got it marked as required in the provider yaml too)